### ISSUE TITLE
Query Regarding Accuracy Discrepancy Between Static and Dynamic Sampling

### ISSUE THREAD

**Issue by Crys-Chen** [Gen-Verse/dLLM-RL#5]
Hi there,

First, thank you for the excellent work on this project.

I've observed a non-trivial accuracy difference between the static sampling and dynamic sampling methods in your experimental results.

This is particularly interesting because, as you know, the [Fast-dLLM paper](https://arxiv.org/abs/2505.22618) not only mathematically proves the equivalence of these two approaches but also demonstrates the stability of dynamic sampling in their own experiments.

Given this, I was hoping you could share any insights into what might be causing this discrepancy in your implementation. For instance, could it be related to specific numerical precision issues, hardware configurations, or other implementation details?

<img width="1071" height="481" alt="Image" src="https://github.com/user-attachments/assets/cc6723ab-f6b6-4910-bb39-e660972d6716" />

Thanks for any insights you can provide!

**Comment by Labman42** [comment]
> Thanks for the quick and insightful reply! That context is really helpful.
> 
> Your point about the theoretical assumptions and over-confidence is well-taken. It seems plausible that the issue lies in how the confidence metric is defined and calculated differently depending on the model architecture and training methods.
> 
> Out of curiosity, were the models trained using only static sampling? If that's the case, I wonder if fine-tuning or retraining them with dynamic sampling might help close the performance gap.
> 
> On a separate note, I also ran into an issue while evaluating the newer [SDAR models](https://huggingface.co/JetLM). I noticed that while several block sizes are available, the evaluation fails on any model with a block size greater than 16.
> 
> <img alt="Image" width="916" height="415" src="https://private-user-images.githubusercontent.com/105596152/489024395-63587b25-834b-4dc6-a33c-e2488d0907a5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg0MTI3OTIsIm5iZiI6MTc1ODQxMjQ5MiwicGF0aCI6Ii8xMDU1OTYxNTIvNDg5MDI0Mzk1LTYzNTg3YjI1LTgzNGItNGRjNi1hMzNjLWUyNDg4ZDA5MDdhNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyMFQyMzU0NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMjliMDBmNmMzZmMwZWZhNGYwNDdjNjIxZmRlNmUzYjY3ODg5NjZhNTVlNTY2YjBlNzNjNjQ2YTMyZWYzMzcxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.1q0iRF_QDC17Utc8pHISd3M2xFIeCWcrlqTD05R510o">
> It appears the root cause is the older, pinned version of the jetengine dependency. I tried to resolve this by replacing the directory with the latest version from the jetengine repo. While this stopped the initial error, the models with block sizes > 16 then seemed to get stuck in an infinite loop.
> 
> Would you be able to look into this when you have a chance? Please let me know if I can provide more details or help test a potential fix.

Hi, I am the core author of the SDAR and the developer of JetEngine project. I just noticed that you encountered the issue that the jetengine can not properly running the model with block size > 16.  I tried your issue with 1.7B SDAR with block size 32 and 64, but I did not reproduced your problem. Is it possible that you can reproduce the issue and give me more detialed information? And it would be better that you can describe your issue in the SDAR issue or JetEngine issue for us to address.
> Out of curiosity, were the models trained using only static sampling? 
The larger block size of all the SDAR series model is not achieved by RL.

**Comment by Crys-Chen** [comment]
Thanks for looking into this!

I was using the 8B SDAR models with a block size larger than 16, and it seems the problem stems from the outdated version of jetengine used by dLLM-RL.

I followed the suggestion to set a smaller max_active value, which has resolved the issue for now.

For a more permanent solution, I'd recommend you to updating the jetengine dependency to the latest version for dLLM-RL. This might resolve the underlying problem.

**Comment by Crys-Chen** [comment]
> Hi, I am the core author of the SDAR and the developer of JetEngine project. I just noticed that you encountered the issue that the jetengine can not properly running the model with block size > 16. I tried your issue with 1.7B SDAR with block size 32 and 64, but I did not reproduced your problem. Is it possible that you can reproduce the issue and give me more detialed information? And it would be better that you can describe your issue in the SDAR issue or JetEngine issue for us to address.

Thanks for looking into this!

I was using the 8B SDAR models with a block size larger than 16, and it seems the problem stems from the outdated version of jetengine used by dLLM-RL.

I followed the suggestion to set a smaller max_active value, which has resolved the issue for now.

For a more permanent solution, I'd recommend you to updating the jetengine dependency to the latest version for dLLM-RL. This might resolve the underlying problem.

**Comment by yinjjiew** [comment]
Thank you guys. I will also keep notice of jetengine's update and update this repo~

**Comment by Labman42** [comment]
Test with SDAR-8b-Chat-b32, still can not reproduce your problem. And it would be better to use the lastest version of flashattn and triton with jetengine. We've tested jetengine on triton >= 3.2.0 and torch >= 2.6.0

**Comment by Crys-Chen** [comment]
I run the models on H100 GPUs with torch 2.8.0 and CUDA 12.8. Perhaps you could reach out to the dLLM-RL authors to see why they recommended to use a smaller `max_active` value like 16 for block size of 64. I believe they would have the best insight into this issue.

**Comment by cychomatica** [comment]
I think the significant accuracy drop brought by dynamic sampling is likely to be attributed to the temperature. In the [Fast-dLLM paper](https://arxiv.org/abs/2505.22618), they did not explicitly explain the temperature they used. But from their repo, they use a temperature of 0 by default, which is just greedy decoding by argmaxing the raw softmax probs. However, in Trace-RL, the temperature is set to 0.1 for Dream by default, which somehow brings overconfidence to some tokens where the raw softmax probs do not reach the threshold and is equivalent to lowering the threshold. 

I tried some different temperatures $\tau$ with a default threshold of 0.95 for Dream on HumanEval. When using a block size of 32:

| Static | Dynamic $\tau$=0 | Dynamic $\tau$=0.1 | Dynamic $\tau$=1 | 
| - | - | - | - |
| 55.49 | 55.49 | 1.22 | 46.95 |

When using a block size of 4, the accuracy drop is not that catastrophic, as the maximum overconfident tokens decoded per step are 4:

| Static | Dynamic $\tau$=0 | Dynamic $\tau$=0.1 | Dynamic $\tau$=1 | 
| - | - | - | - |
| 56.10 | 56.10 | 28.66 | 27.44 |
